### PR TITLE
Remove compile bottlenecks from ZImage pipeline

### PR DESCRIPTION
**Pull request by hitchhiker3010**
# What does this PR do?

Fixes performance issues identified by profiling `ZImagePipeline` with `torch.profiler` as part of #13401 .

### What does this PR do?

Profiled `ZImagePipeline` (using `Tongyi-MAI/Z-Image-Turbo`) in both eager and `torch.compile` modes following the [profiling guide](https://github.com/huggingface/diffusers/tree/main/examples/profiling). The Chrome traces revealed two device-to-host (DtoH) synchronization points that break asynchronous GPU execution and prevent `torch.compile` from yielding its full speedup.

**Pipeline denoising loop: `t_norm = timestep[0].item()` DtoH sync**

1. Inside the denoising loop, `timestep[0].item()` triggers a GPU→CPU sync every step to read `t_norm` for CFG truncation logic. Since the full timestep schedule is known before the loop begins, we precompute all `t_norm` values into a plain Python list *before* entering the loop and index into it with `i`. 
2. This also lets us set `scheduler.set_begin_index(0)` upfront to avoid the DtoH sync in `_init_step_index` (same pattern as #11696 ) 

Profiling ZImagePipeline 
GPU - L4
num_inference_steps - 4,
guidance_scale - 0.0 ( Guidance should be 0 for the Turbo models)

Before 
image
The first scheduler_step took 657.8µs
Number of cudaStreamSynchronize blocks - 19

After 
image
The first scheduler_step took 15.49 µs after this fix
Number of cudaStreamSynchronize blocks - 13
Part of #13401 .

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case. [Here is the link to the discussion](https://github.com/huggingface/diffusers/issues/13401#issuecomment-4205856594)
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sayakpaul @dg845

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for your PR! Can we eliminate all the `cudaStreamSynchronize` calls?

**Comment by hitchhiker3010** [comment]
Here are some comparison stats between [commit_1](https://github.com/huggingface/diffusers/pull/13461/commits/b04b66f36d9e2dc906667ec1606f0842cbcfa7fa) and [commit_2](https://github.com/huggingface/diffusers/pull/13461/commits/2da67248e0d4c2cda112ed9e962dbcee90dffa08)

Metric | commit_1 eager | commit_2 eager | commit_1 compile | commit_2 compile
-- | -- | -- | -- | --
nonzero calls | 28 | 4  | 28 | 4 
nonzero CPU time | 2091 ms | 0.72 ms | 3057 ms | 0.49 ms
index_put calls | 20 | 0 | 36 | 0
index_put total | 4183 ms | 0 ms | 9172 ms | 0 ms
cudaStreamSynchronize calls | 13 | 5  | 13 | 5 
cudaStreamSynchronize total | 2089 ms | 0.47 ms | 3055 ms | 0.32 ms

**Comment by hitchhiker3010** [comment]
all the trace files can be [accessed here.](https://drive.google.com/drive/folders/1-aR-RrnOJ6vz1Fg3BIEg1xg4sv8P3d0g?usp=share_link)

The `cudaStreamSynchronize` traces from the Denoising phase are eliminated now, the remaining 5 `cudaStreamSynchronize` seem to be from the text encoding phase, should we fix them too?

cc: @sayakpaul

**Review by dg845 — APPROVED** [review]
Thanks for the PR!

**Comment by sayakpaul** [comment]
Merging as the outputs with and without this PR are the same:

```py
"""
Minimal script to verify PR #13461 does not change ZImagePipeline outputs.

Compares latent outputs between the current branch (hitchhiker3010-main, with PR changes)
and the main branch (without PR changes) using a fixed seed.

The PR makes two changes:
  1. Precompute cfg_truncation t_norms outside the loop (avoids DtoH sync)
  2. Use torch.where instead of boolean mask indexing in the transformer

Both are pure optimizations — outputs should be identical.

Usage:
    # On the current branch (with PR changes):
    python test_zimage_pr13461.py --save current_branch.pt

    # On the main branch (without PR changes):
    git checkout main
    python test_zimage_pr13461.py --save main_branch.pt

    # Compare:
    python test_zimage_pr13461.py --compare current_branch.pt main_branch.pt
"""

import argparse
import torch


def run_pipeline():
    from diffusers import ZImagePipeline

    pipe = ZImagePipeline.from_pretrained(
        "Tongyi-MAI/Z-Image-Turbo",
        torch_dtype=torch.bfloat16,
    )
    pipe.to("cuda")

    generator = torch.Generator(device="cuda").manual_seed(0)

    # Use guidance_scale > 1 and cfg_truncation to exercise both changed code paths.
    # Small resolution + few steps + latent output for speed.
    output = pipe(
        prompt="a cat",
        height=256,
        width=256,
        num_inference_steps=2,
        guidance_scale=3.5,
        cfg_truncation=0.5,
        output_type="latent",
        generator=generator,
    )
    return output.images


def save(latents, path):
    torch.save(latents.cpu(), path)
    print(f"Saved latents with shape {latents.shape} and dtype {latents.dtype} to {path}")


def compare(path_a, path_b):
    a = torch.load(path_a, weights_only=True)
    b = torch.load(path_b, weights_only=True)

    print(f"Tensor A: shape={a.shape}, dtype={a.dtype}")
    print(f"Tensor B: shape={b.shape}, dtype={b.dtype}")

    if a.shape != b.shape:
        print("FAIL: shapes differ")
        return

    exact_match = torch.equal(a, b)
    max_diff = (a.float() - b.float()).abs().max().item()
    print(f"Exact match: {exact_match}")
    print(f"Max absolute difference: {max_diff}")

    if exact_match:
        print("PASS: outputs are identical")
    elif max_diff < 1e-3:
        print(f"PASS: outputs differ by at most {max_diff} (within tolerance)")
    else:
        print(f"FAIL: outputs differ by {max_diff}")


def main():
    parser = argparse.ArgumentParser(description="ZImage PR #13461 output comparison")
    parser.add_argument("--save", type=str, help="Run pipeline and save latents to this path")
    parser.add_argument("--compare", nargs=2, metavar=("A", "B"), help="Compare two saved latent files")
    args = parser.parse_args()

    if args.save:
        latents = run_pipeline()
        save(latents, args.save)
    elif args.compare:
        compare(args.compare[0], args.compare[1])
    else:
        parser.print_help()


if __name__ == "__main__":
    main()

```

**Comment by hitchhiker3010** [comment]
Hey @sayakpaul @dg845 

Thanks for the contributing opportunity, there must be similar issues in the other Z Image pipelines[controlnet, controlnet_inpaint, img2img, inpaint and omni] along with some pipeline specific ones, can I look into those or do you suggest any other pipelines that might be of priority?
